semantics/src/main/scala/aqua/semantics/rules/locations/LocationsState.scala

```diff
@@ -1,29 +1,60 @@
 package aqua.semantics.rules.locations
 
 import aqua.parser.lexer.Token
-import aqua.semantics.rules.types.TypesState
 
 import cats.kernel.Monoid
+import scribe.Logging
 
 case class LocationsState[S[_]](
-  tokens: List[(String, ExprInfo[S])] = Nil,
-  locations: List[(Token[S], Token[S])] = Nil
-) {
+  variables: Map[Token[S], VariableInfo[S]] = Map.empty[Token[S], VariableInfo[S]],
+  nameToVariableList: List[(String, Token[S])] = Nil
+) extends Logging {
 
-  lazy val allLocations: List[(Token[S], Token[S])] = locations
+  private def insertIfNotExists(
+    vars: Map[Token[S], VariableInfo[S]],
+    definition: DefinitionInfo[S]
+  ): Map[Token[S], VariableInfo[S]] = {
+    vars + vars
+      .get(definition.token)
+      .map(definition.token -> _)
+      .getOrElse(definition.token -> VariableInfo(definition))
+  }
```

Comment:
I think it would be shorter to use `vars.updateWith(_.getOrElse(VariableInfo(definition)).some)`